File: shell_sort/shell_sort.cpp
#include <iostream>

using namespace std;

void printArray(int arr[], int n);
int shellSort(int arr[], int n);

int main()
{
    int sizeArr;
    cout<<"Enter Array Size:"<<endl;
    cin>>sizeArr;
    int arr[sizeArr];

    for(int i=0;i<sizeArr;i++)
    {
        cout<<"Enter Value for array["<<i<<"]"<<endl;
        cin>>arr[i];

    }


    shellSort(arr, sizeArr);
    cout<<"Array after sorting:"<<endl;
    printArray(arr, sizeArr);
    return 0;
}


void printArray(int arr[], int n)
{
    for (int i=0; i<n; i++)
        cout<<arr[i]<<endl;
}

int shellSort(int arr[], int n)
{

    for (int gap = n/2; gap > 0; gap /= 2)
    {
        for (int i = gap; i < n; i += 1)
        {
            int temp = arr[i];
            int j;
            for (j = i; j >= gap && arr[j - gap] > temp; j -= gap)
                arr[j] = arr[j - gap];
            arr[j] = temp;
        }
    }
    return 0;
}

File: algorithms/ar-shsrt/python3/shell_sort.py
def shellSort(arr):    
    n = len(arr)     
    # Rearrange elements at each n/2, n/4, n/8, ... gaps
    gap = n//2  
    while gap > 0:   
        for i in range(gap,n):             
            temp = arr[i] 
            j = i 
            while  j >= gap and arr[j-gap] >temp: 
                arr[j] = arr[j-gap] 
                j -= gap   
            arr[j] = temp 
        gap //= 2

def main():
    arr = [ 1,31,78,26,11,16,46] 
    shellSort(arr) 
    print(arr) 

if __name__ == '__main__':
    main() 